### DISCUSSION TITLE
In what scenarios is FastAPI particularly good at compared to WSGI?

### DISCUSSION THREAD

**Discussion by pwtail** [fastapi/fastapi#5945] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
def application(env, start_response):
    start_response('200 OK', [('Content-Type', 'text/html')])
    yield "WSGI is a dinosaur"
```

### Description

Hi, I a team lead, and our project is of the e-shop kind more or less. Pretty loaded - up to 5000 purchases daily. The performance is overall OK but we can't always handle the load at pikes (half of the reasons for that being bad code of course). 

The backend team are Python devs, including myself. The team is pretty  enthusiastic about FastAPI, and we currently consider using it for our new services. **What scenarios we can expect performance wins at?**

We currently use flask and sqlalchemy. Postgres as database.

Some endpoints make requests to another services via http. However, there are definitely less than 50% of those. Currently just using `requests` library for that. In our setup, the number of Python threads are higher than the number of cores, so we think it should help solve the "requests" problem.

We use a separate async service currently for sending websocket notifications, so one thing that we know for sure is going to improve is working with websockets.

We have a lot of heavy db queries, the majority however being light and medium ones.

What are the scenarios that WSGI particularly sucks at compared with FastAPI? 

Currently we use uwsgi for deploy. It doesn't have ASGI integration so will need something else if we decide in favor of FastAPI.

### Operating System

Linux

### FastAPI Version

0.89.1

### Python Version

Python 3.10.9

**Comment by jgould22** [comment, marked as answer]
WSGI is a specification, FastAPI is a library, they are not super comparable but I get what you mean, sync vs async.

Some things to think about

> 5000 purchases daily 

Is not really a meaningful metric for us to base on. From the outside looking in that seems like a pretty small number but I do not know what is required for each purchase.

> What scenarios we can expect performance wins at?

The answer may be none. If you flask app is provisioned adequately they may preform the same. 

Async is often about "doing more with less".  

So you may see a reduction in resource usage which can manifest as a performance improvement if your flask app is under provisioned. 

> Some endpoints make requests to another services via http. However, there are definitely less than 50% of those. Currently just using requests library for that. In our setup, the number of Python threads are higher than the number of cores, so we think it should help solve the "requests" problem.

There are equivalent async libraries that can preform requests with similar apis.

> We have a lot of heavy db queries, the majority however being light and medium ones.

This is talking in two directions, regardless if you are waiting a lot on the database then you will see minimal difference between the two (since all the work is happening in the database)

> What are the scenarios that WSGI particularly sucks at compared with FastAPI?

Generally async code is better if you have a lot of connections that are waiting around on something external, ie database calls or external services.

**Comment by pwtail** [comment]
Thank you for sharing your opinion. You aren't a natural born sales person, are you?) Well, I think I can agree the advantages of async may be minimal in my case.

**Reply by jgould22** [reply]
I've spent a lot of a time converting older sync code bases to async (for improved resource usage as opposed to speed) so I have a lot of experience with the rougher edges of the ecosystem. 

FastAPI has some nice features that make it worth it (the pydantic integration is very useful) but it is not inherently "faster" than flask and can be slower if one of your dependencies is sync. 

Best of luck with you app.
